File: lc572_subtreeOfTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSametree(self, s: TreeNode, t: TreeNode) -> bool:
        if s==None and t==None:
            return True
        elif s==None and t!=None:
            return False
        elif s!=None and t==None:
            return False
        else:
            if s.val == t.val:
                return self.isSametree(s.left, t.left) and self.isSametree(s.right, t.right)
            else:
                return False
            
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        if self.isSametree(s, t):
            return 1
        else:
            if s != None:
                return self.isSubtree(s.left, t) or self.isSubtree(s.right, t)
            else:
                return False

File: lc572_subtreeOftree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool issame(struct TreeNode* s, struct TreeNode* t){
    if (s==NULL && t== NULL)return 1;
    if (s==NULL&&  t!= NULL)return 0;
    if (s!=NULL&&  t== NULL)return 0;
    if(s->val == t->val){
        return (issame(s->left, t->left) && issame(s->right, t->right));
    }else{
        return 0;
    }
}
bool isSubtree(struct TreeNode* s, struct TreeNode* t){
    
    if (issame(s, t)){
        return 1;
    }else{
        if(s!=NULL)
            return (isSubtree(s->left, t) || isSubtree(s->right, t));
        else
            return 0;
    }
}
